### PR TITLE
Write the listing-change and race journeys as stories

### PR DESCRIPTION
**Pull request by stefan-burke**
## What changed

More multi-day booking tests rewritten as stories.

**A new story — a listing changes after people have already booked**

- The organiser stops opening on a day of the week that a booked stay already covers. The stay keeps all three of its days, and only new stays are kept off that day.
- The organiser makes stays longer than the window the listing takes bookings for. The stay already sold keeps its stretched days, but the day it starts on stops being offered to anyone new.

**Two rules added to the multi-day booking story**

- **Longer stays leave fewer days to start on.** A stay has to finish inside the booking window, so a five-day listing offers fewer start days than a one-day one — but still offers some. This merges two near-identical tests that lived in separate files.
- **Two customers racing for the last stay get one booking between them.** Both press Continue at the same moment; exactly one is taken and the listing ends up holding one stay.

## Something the migration turned up

One of the tests being migrated was about a listing that starts being booked by the day when it was booked as one date before. Written out as a story — "people booked before the change still take up room" — it failed, because that is not what the site does.

The people who booked before the switch have no day of their own, and the per-day count deliberately skips them. So a listing with both its places taken stops being full the moment it is switched, and can end up holding three people where it has room for two. The listing's own total check still counts them, so the two checks disagree.

That may well be intended, but nothing says so, and an organiser switching a sold-out listing would not expect it to reopen. It is outside a migration whose job is to move existing tests, so it is written up in `TODO.md` with the three ways it could be settled. The story asserting the behaviour was removed rather than left passing on a claim the product does not make, and the original test stays where it is.

## What stayed a direct test

The checks of the group-limit reconciler and the capacity helpers are technical contracts, not journeys, so they stay as direct tests — a story may never be the only thing covering a line of the product.

## Checks

`deno task precommit` passes in full — typecheck, lint, duplication at 0%, the whole test suite, and 100% line and branch coverage. All 70 Cucumber scenarios (463 steps) pass. No production code changed.

---
_Generated by [Claude Code](https://claude.ai/code/session_01Ga3qdj1PzXyZTcjmXZRqz7)_

## Summary by CodeRabbit

- **Booking Availability**
  - Added coverage for how booking windows affect available start dates, especially for longer stays.
  - Prevented overbooking when multiple customers attempt to reserve the final available stay simultaneously.

- **Listing Changes**
  - Existing bookings remain valid when available weekdays or booking windows are changed.
  - New bookings are restricted according to the updated listing rules.

- **Documentation**
  - Recorded an open product decision regarding undated bookings when switching to daily booking.

### CODE REVIEW

File: test/specs/support/listing-changes.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+/** The organiser stops opening a listing on one day of the week. */
+export const stopOpeningOn = async (
+  world: TicketsWorld,
+  name: string,
+  weekday: string,
+): Promise<void> => {
+  await updateTestListing(stayListing(world, name).id, {
+    bookableDays: WEEKDAYS.filter((day) => day !== weekday),
+  });
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/specs/support/listing-changes.ts]: Fair — fixed in `9f9af44d`. `stopOpeningOn` now opens the listing's edit page in the story browser, reads the ticked `bookable_days` entries off the served HTML with `extractFormEntries`, asserts the day it means to untick is actually offered, and submits the remaining days through the page's own form (asserting the "Listing updated" confirmation). A page that stops rendering the checkboxes now fails the story rather than passing on a rebuilt POST.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: test/specs/steps/listing-changes.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    name: string,
+    startsIn: number,
+  ): Promise<void> {
+    const listing = stayListing(this, name);
+    const day = dayFromToday(this, startsIn);
+    // Both press Continue together, so the site has to settle the race itself.
+    const attempts = await Promise.all([
+      visitorTriesToBook(listing, { ...guest(1), day }),
+      visitorTriesToBook(listing, { ...guest(2), day }),
+    ]);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/specs/steps/listing-changes.ts]: Correct, and I should have kept it — a story may never be the only cover of a branch, and the remaining atomic-create race is undated so it never reaches the range check. Restored in `9f9af44d` as a direct test in `test/e2e/duration-days/edge-cases.test.ts` ("two simultaneous multi-day bookings for one place — only one is taken"), with a header comment naming the story it sits beside and why both exist.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: test/specs/steps/listing-changes.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    this.raceWinners = attempts.filter(({ wasBooked }) => wasBooked).length;
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/specs/steps/listing-changes.ts:104]: Agreed — fixed in `9f9af44d`. The When keeps the losing `BookingAttempt`, and the Then runs it through `expectRefusedForWantOfRoom` before checking the row count, so a validation or server error on one request no longer reads as the site settling the race.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: test/specs/support/listing-changes.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+import { stayListing } from "#test/specs/support/stays.ts";
+import type { TicketsWorld } from "#test/specs/support/world.ts";
+import { updateTestListing } from "#test-utils/db-helpers/listings.ts";
+
+const WEEKDAYS = [
+  "Sunday",
+  "Monday",
+  "Tuesday",
+  "Wednesday",
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/specs/support/listing-changes.ts]: Fixed in `9f9af44d` — the helper now imports `DAY_NAMES` from `#shared/day-names.ts`. It suits this use better than `VALID_DAY_NAMES` because it runs Sunday-first, the same order a date reports its weekday in, so the lookup is a direct index with no reordering.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: specs/bookings/changes-after-people-have-booked.feature [hunk 5]
```diff
@@ -0,0 +1,38 @@
+@story:bookings.changes-after-people-have-booked
+@owner:bookings @risk:high
+@actor:organiser @actor:customer
+@edition:managed @edition:self-hosted
+Feature: A listing changes after people have already booked
+  An organiser can close a day of the week, make stays longer, or start booking
+  a listing by the day when it was not before. Whatever they change, the people
+  who already booked keep what they were sold — and the room they take up is
+  still counted.
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply specs/bookings/changes-after-people-have-booked.feature]: You're right, and this is the sharpest of the five — the description would have published a rule the catalog never checks, and one the product does not follow. Fixed in `9f9af44d`: the narrative now covers only what the two Rules exercise (closing a weekday, and stretching stays past the booking window), and the type-switch and blanket capacity claim are gone. The open question stays in `TODO.md` until it is decided.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: specs/bookings/changes-after-people-have-booked.feature [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  Rule: A stay stretched past how far ahead people can book is still kept
+    Making stays longer can push one already booked past the last day the
+    listing takes bookings for. That stay was sold in good faith, so it keeps
+    its days — but no new stay can start there.
+
+    @case:changes.a-stay-stretched-past-the-window
+    Scenario: The organiser makes stays longer than the booking window allows
+      Given a Cabin that takes bookings 10 days ahead, 1 day at a time
+      And a customer booked a Cabin stay starting in 9 days
+      When the organiser makes each Cabin stay 5 days long
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply specs/bookings/changes-after-people-have-booked.feature:35]: Right, and it applied to every story using that helper, not just this scenario. Fixed in `1047ba90`: `changeStayLength` now runs the new length through `whyValueCannotBeSent` — the same check the public booking journeys use — so a duration box that is missing, disabled, or hidden at another value fails the story instead of the post being forced through.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: test/specs/steps/listing-changes.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    name: string,
+    startsIn: number,
+  ): Promise<void> {
+    const listing = stayListing(this, name);
+    const day = dayFromToday(this, startsIn);
+    // Both press Continue together, so the site has to settle the race itself.
+    const attempts = await Promise.all([
+      visitorTriesToBook(listing, { ...guest(1), day }),
+      visitorTriesToBook(listing, { ...guest(2), day }),
+    ]);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/specs/steps/listing-changes.ts]: Agreed — the scenario says "at the same moment" and `Promise.all` over two GET-then-POST sequences did not guarantee it. Fixed in `1047ba90` with a barrier: `visitorFillsInOrder` now opens the page and checks every control, then returns a `press()` for the submit. The race step fills both forms in first and only then awaits both `press()` calls, so neither can complete before the other has loaded. `visitorTriesToOrder` is the same thing pressed straight away, so every other story is unchanged.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: specs/bookings/changes-after-people-have-booked.feature [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+      Then the organiser sees that stay now runs for 5 days
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply specs/bookings/changes-after-people-have-booked.feature:36]: Valid, and it was the circular-expectation trap I had just written into `E2E_TESTS.md` sitting in a helper I did not re-read. Fixed in `4f992583`: `expectStayRunsFor` now checks the days the site *stored* on the booking against the scenario's own first day plus its own number of days — arithmetic `formatDateRangeLabel` plays no part in — and keeps the rendered-label check on top, so the organiser is still proven to see it. The stored-end assertion the old test made is back, and a miscalculated label can no longer pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: test/specs/support/stays.ts [hunk 9]
```diff
@@ -112,8 +120,14 @@ export const changeStayLength = async (
...
   await browser.visit(`/admin/listing/${stayListing(world, name).id}/edit`);
-  expect(browser.currentHtml).toContain('name="duration_days"');
-  await browser.submitForm({ duration_days: String(days) }, "Save Changes");
+  const length = String(days);
+  // The organiser has to be able to send this length: a box that is missing,
+  // disabled, or fixed at another value would mean they cannot make the change
+  // at all, however happily the form post is accepted.
+  expect(
+    whyValueCannotBeSent(browser.currentHtml, "duration_days", length),
+  ).toBeNull();
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/specs/support/stays.ts:129]: Correct — the helper had the gap, not just this call site. Fixed in `d6155a6e`: `whyValueCannotBeSent` now refuses a `readonly` box, and refuses a number outside a box's own `min`/`max` ("the duration_days box takes nothing above 3"). Non-numeric and empty values have no range to break, so text boxes are unaffected. Six direct tests in `form-controls.test.ts` cover the new rules, and every story that fills in a form gets them for free.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: test/specs/support/listing-changes.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+export const stopOpeningOn = async (
+  world: TicketsWorld,
+  name: string,
+  weekday: string,
+): Promise<void> => {
+  const browser = await adminBrowser(world);
+  await browser.visit(`/admin/listing/${stayListing(world, name).id}/edit`);
+  const ticked = extractFormEntries(browser.currentHtml)
+    .filter(([field]) => field === "bookable_days")
+    .map(([, day]) => day);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/specs/support/listing-changes.ts]: Right — `extractFormEntries` counts a hidden input as a successful control, so a day nobody could untick would have read as ticked. Fixed in `d6155a6e` with a new pure helper, `tickedCheckboxes(html, field)`, which counts a value only when it comes from an `<input type="checkbox">` for that field that is `checked` and not `disabled`. `stopOpeningOn` uses it, so a day carried by a hidden box now fails the story at the "is it ticked" assertion. Three direct tests cover the hidden, disabled, and other-field cases.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: test/specs/steps/stays-booking.ts [hunk 11]
```diff
@@ -195,24 +206,36 @@ When(
...
 
-/** What the organiser reads on the booking's own page. The label must name the
- * stay's first and last day, so a stay stored a day short or long fails. */
+/** How long the stay runs, checked two ways that cannot both drift together.
+ *
+ * The days the site stored are compared against the day the story started on
+ * plus its own number of days — arithmetic the site's own wording plays no part
+ * in. Then the booking's page is read, so the organiser is shown a label naming
+ * exactly those days. Checking only the label would let a change to how ranges
+ * are worded pass unnoticed, because the expected label is built by the same
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/specs/steps/stays-booking.ts]: Fair — eight lines re-narrating four. Cut to three in `d6155a6e`, keeping only the part the code cannot say: the label alone would not do, because it is built by the same helper the page renders with.

---
_Generated by [Claude Code](https://claude.ai/code)_
